### PR TITLE
Initial integration of ModelingToolkit

### PR DESCRIPTION
**Pull request by yashvardhan747**

### CODE REVIEW

**Comment by dpsanders** [comment]
It no longer seems to be possible to call a contractor:
```
julia> C(-Inf..1, IntervalBox(-10..10, 2))
ERROR: MethodError: objects of type IntervalConstraintProgramming.GeneratedFunction{getfield(IntervalConstraintProgramming, Symbol("##47#48"))} are not callable
Stacktrace:
 [1] (::Contractor{2,1,getfield(IntervalConstraintProgramming, Symbol("##47#48")),getfield(IntervalConstraintProgramming, Symbol("##49#50")),Operation})(::IntervalBox{1,Float64}, ::IntervalBox{2,Float64}) at /Users/dpsanders/.julia/dev/IntervalConstraintProgramming/src/contractor.jl:47
 [2] (::Contractor{2,1,getfield(IntervalConstraintProgramming, Symbol("##47#48")),getfield(IntervalConstraintProgramming, Symbol("##49#50")),Operation})(::Interval{Float64}, ::IntervalBox{2,Float64}) at /Users/dpsanders/.julia/dev/IntervalConstraintProgramming/src/contractor.jl:71
 [3] top-level scope at none:0
```

File: src/code_generation.jl [hunk 1]
```diff
@@ -10,7 +10,7 @@ end
 
 # GeneratedFunction(code::Expr) = GeneratedFunction(eval(code), code)
 
-(f::GeneratedFunction{F})(x...) where {F} = f.f(x...)
+#(f::GeneratedFunction{F})(x...) where {F} = f.f(x...)
```
Comment (dpsanders) [inline src/code_generation.jl]: Why did you comment out this line?

File: src/contractor.jl [hunk 2]
```diff
@@ -70,6 +70,34 @@ end
 
 (C::Contractor{N,1,F1,F2})(A::Interval{T}, X::IntervalBox{N,T}) where {N,F1,F2,T} = C(IntervalBox(A), X)
```
Comment (dpsanders) [inline src/contractor.jl]: Maybe you need to add the `ex` type parameter in order to make this callable? (I'm not sure)

File: src/code_generation.jl (hunk 1 above)
Comment (dpsanders) [reply src/code_generation.jl]: If you uncomment this again then you can call `Contractor`

**Comment by dpsanders** [comment]
The `gensym` stuff is not working - the following should use gensymmed variables in the generated code.
```
julia> f(x) = x + x + x + x + x + x + x + x +  x + x + x + x +  x + x + x + x +  x + x + x + x +  x + x + x + x +  x + x + x + x +  x + x + x + x 

julia> C = Contractor(f(x))

julia> C.forward.code
quote
    #= /Users/dpsanders/.julia/dev/IntervalConstraintProgramming/src/code_generation.jl:193 =#
    t->begin
            #= /Users/dpsanders/.julia/dev/IntervalConstraintProgramming/src/code_generation.jl:193 =#
            #= /Users/dpsanders/.julia/dev/IntervalConstraintProgramming/src/code_generation.jl:194 =#
            () = t
            #= /Users/dpsanders/.julia/dev/IntervalConstraintProgramming/src/code_generation.jl:195 =#
            begin
                r = x + x
                s = r + x
                t = s + x
                u = t + x
                v = u + x
                w = v + x
                x = w + x
                y = x + x
                z = y + x
                a = z + x
                b = a + x
                c = b + x
                d = c + x
                e = d + x
                f = e + x
                g = f + x
                h = g + x
                i = h + x
                j = i + x
                k = j + x
                l = k + x
                m = l + x
                n = m + x
                o = n + x
                p = o + x
                q = p + x
                r = q + x
                s = r + x
                t = s + x
                u = t + x
                v = u + x
            end
            #= /Users/dpsanders/.julia/dev/IntervalConstraintProgramming/src/code_generation.jl:196 =#
            return (v, (r, s, t, u, w, x, y, z, a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q))
        end
end
```

File: src/ast.jl [hunk 3]
```diff
@@ -1,18 +1,7 @@
...
-
-    if i == 0
-        return Symbol("_", s)
-    else
-        return Symbol("_", s, i)
-    end
-end
+"""Return a new symbol"""
 
+make_symbol(s::Symbol)= Symbol(s)
```
Comment (dpsanders) [inline src/ast.jl]: What happened to `gensym`? We need it.

File: src/code_generation.jl (hunk 1 above)
Comment (yashvardhan747) [reply src/code_generation.jl]: Yes thats by mistake.

**Review by yashvardhan747 — COMMENTED** [review]
Actually It was little hard to read code while using gensym. So I removed it temporarily. I'll correct it.

**Comment by dpsanders** [comment]
> Actually It was little hard to read code while using gensym. So I removed it temporarily. I'll correct it.

I agree, it's harder to read with `gensym`. So let's go back to the previous version that was easier to read!

**Comment by dpsanders** [comment]
You will need to add `ModelingToolkit` in `REQUIRE`.

**Comment by yashcodes** [comment]
> You will need to add `ModelingToolkit` in `REQUIRE`.

**Comment by yashcodes** [comment]
I already added it in IntervalConstraintProgramming.jl

**Comment by yashcodes** [comment]
where else I need to add ModelingToolkit

**Comment by dpsanders** [comment]
Thanks a lot!

**Pull request merged**